### ISSUE TITLE
Shader discards structure member uniforms.

### ISSUE THREAD

**Issue by rozenmad** [love2d/love#1706]
Here's an example:

```lua
local shader = love.graphics.newShader([[
struct TestStruct {
      vec4 color;
};
uniform TestStruct test[1];

#ifdef PIXEL
      vec4 get_color(in TestStruct test) {
            return test.color;
      }
      vec4 effect(vec4 color, Image t, vec2 texture_coords, vec2 screen_coords)
      {
            //test[0].color;
            return get_color(test[0]);
      }
#endif
]])

function love.load()
      shader:send('test[0].color', {0.2, 0.6, 1, 1})
end
function love.draw()
      love.graphics.setShader(shader)
      love.graphics.rectangle('fill', 0, 0, 100, 100)
      love.graphics.setShader()
end
```

Error
main.lua:20: Shader uniform 'test[0].color' does not exist.
A common error is to define but not use the variable.

```lua

local shader = love.graphics.newShader([[
struct TestStruct {
      vec4 color;
};
uniform TestStruct test;

#ifdef PIXEL
      vec4 get_color(in TestStruct t) {
            return t.color;
      }
      vec4 effect(vec4 color, Image t, vec2 texture_coords, vec2 screen_coords)
      {
            //test.color;
            return get_color(test);
      }
#endif
]])

function love.load()
      shader:send('test.color', {0.2, 0.6, 1, 1})
end
function love.draw()
      love.graphics.setShader(shader)
      love.graphics.rectangle('fill', 0, 0, 100, 100)
      love.graphics.setShader()
end
```
Error
main.lua:22: Shader uniform 'test.color' does not exist.
A common error is to define but not use the variable.

If you uncomment the line in shaders, then everything works. Since there is an explicit reference to a member of the structure of the uniform.

**Comment by slime73** [comment]
The code works on my computer  – this looks like a graphics driver bug, and unfortunately LÖVE's source code can't do anything to fix your graphics driver optimizing out variables it shouldn't.

One thing you could try is updating your driver if possible.

**Comment by rozenmad** [comment]
This is probably a driver error. I have several devices with Intel graphics, such as Surface Pro 3 with Intel HD 5000, and Pro 2 with HD 4000 (driver dated 06/08/2020), both examples do not work on them. But what's interesting is that there is also an older laptop with HD 3000 on which the second example works.
Everything works fine on the Nvidia GT630 and on my mobile device.

**Comment by slime73** [comment]
According to https://downloadcenter.intel.com/product/81499/Intel-HD-Graphics-4000 there's an Intel HD 4000 driver from October 2020, and at https://downloadcenter.intel.com/download/30196/Intel-Graphics-Driver-for-Windows-15-40- there's an HD 5000 driver from February 2021, maybe you could try those?

**Comment by rozenmad** [comment]
Yes, there are newer drivers on the Intel HD 4000 website, but they cannot be installed, probably they are not compatible with the drivers supplied by Microsoft. I installed them manually, but LOVE doesn't work with them at all. Running the simplest example displays a black screen. Running LOVE without code into demo scene simply doesn't respond.
I have encountered some shader artifacts in other games and have installed these drivers before, this fixed bugs, but some applications stopped working.

I will also try to install new drivers on the HD5000, and I'll write about it later.
